docs/cudf/source/developer_guide/benchmarking.md

```diff
@@ -0,0 +1,323 @@
+def case_1():
+    return 1
+
+def case_2():
+    return 2
+
+def case_3():
+    return 3
+
+# bench_foo.py
+@pytest_cases.parametrize_with_cases(num)
+def bench_foo(benchmark, num):
+    benchmark(num * 2)
+```
+
+`pytest-cases` is allows developers to put complex initialization into named, documented functions.
+That becomes especially valuable when benchmarking APIs whose performance can vary drastically based on parameters.
+Additionally, cases, like fixtures, are lazily evaluated.
+Initializing complex objects inside a `pytest.mark.parametrize` can dramatically slow down test collection,
+or even lead to out of memory issues if too many complex cases are collected.
+Using lazy case functions ensures that the associated objects are only created on an as-needed basis.
+
+When writing cases, just as in writing custom fixtures, developers should make use of the config variables.
+Cases should import the `NUM_ROWS` and/or `NUM_COLS` variables from the config module and use them to define data sizes.
+
+Given the plethora of options for parametrization,
+we codify here some best practices for how each should be employed.
+In general, these approach are applicable to parametrizations of different complexity.
+For the purpose of this discussion,
+we define a "simple parametrization" as parametrization using a list (possibly nested) of primitive objects.
+Examples include a list of integers or a list of list of strings.
+This _does not_ include e.g. cuDF or pandas objects.
+
+With that in mind, here are some ground rules for how to parametrize.
+
+Use `pytest.mark.parametrize` when:
+- One test must be run on many inputs and those inputs are simple to construct.
+
+Use fixtures when:
+- One or more tests must be run on the same set of inputs,
+  and all of those inputs can be constructed with simple parametrizations.
+  In practice, that means that it is acceptable to use a fixture like this:
+  ```python
+      @pytest.fixture(params=['a', 'b'])
+      def foo(request):
+          if request.param == 'a':
+              # Some complex initialization
+          elif request.param == 'b':
+              # Some other complex initialization
+  ```
+  In other words, the construction of the fixture may be complex,
+  as long as the parametrization of that construction is simple.
+
+Use `pytest-cases.parametrize_with_cases` when:
+- One or more tests must be run on the same set of inputs,
+  and at least one of those inputs requires complex parametrizations.
+- Given a set of cases, different tests need to run on different subsets with a nonempty intersection.
+
+
+## Comparing to pandas
+
+An important aspect of benchmarking cuDF is comparing it to pandas.
+We often want to generate quantitative comparisons, so we need to make that as easy as possible.
+Our benchmarks support this by setting the environment variable `CUDF_BENCHMARKS_USE_PANDAS`.
+When this variable is detected, all benchmarks will automatically be run using pandas instead of cuDF.
```

Comment:
Will this also run the `internal/` benchmarks with Pandas? If not, maybe we could say that here.

Reply:
No it will not since we shouldn't expect compatibility with pandas except in public APIs. I can add that note.